services/src/floodlight-modules/src/main/java/org/openkilda/floodlight/service/session/SessionService.java

```diff
@@ -0,0 +1,122 @@
+import org.openkilda.floodlight.service.IService;
+import org.openkilda.floodlight.service.of.IInputTranslator;
+import org.openkilda.floodlight.service.of.InputService;
+
+import com.google.common.annotations.VisibleForTesting;
+import net.floodlightcontroller.core.IOFSwitch;
+import net.floodlightcontroller.core.internal.IOFSwitchService;
+import net.floodlightcontroller.core.module.FloodlightModuleContext;
+import org.projectfloodlight.openflow.protocol.OFMessage;
+import org.projectfloodlight.openflow.protocol.OFType;
+import org.projectfloodlight.openflow.types.DatapathId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class SessionService implements IService, IInputTranslator {
+    private static final Logger log = LoggerFactory.getLogger(SessionService.class);
+
+    private final Map<DatapathId, SwitchSessions> sessionsByDatapath = new ConcurrentHashMap<>();
+
+    /**
+     * Create new OF communication session and register it in service.
+     */
+    public Session open(IOFSwitch sw) {
+        SwitchSessions group;
+        group = sessionsByDatapath.get(sw.getId());
+
+        if (group == null) {
+            throw new IllegalStateException(String.format(
+                    "Switch %s is not registered into %s", sw.getId(), getClass().getName()));
+        }
+
+        return group.open(sw);
+    }
+
+    @Override
+    public void setup(FloodlightModuleContext moduleContext) {
+        InputService inputService = moduleContext.getServiceImpl(InputService.class);
+        inputService.addTranslator(OFType.ERROR, this);
+        inputService.addTranslator(OFType.BARRIER_REPLY, this);
+
+        new SwitchEventsTranslator(this, moduleContext.getServiceImpl(IOFSwitchService.class));
+    }
+
+    @Override
+    public Command makeCommand(CommandContext context, OfInput input) {
+        return new Command(context) {
+            @Override
+            public Command call() throws Exception {
+                handleResponse(input.getDpId(), input.getMessage());
+                return null;
+            }
+        };
+    }
+
+    @VisibleForTesting
+    void handleResponse(DatapathId dpId, OFMessage message) {
+        SwitchSessions group;
+        group = sessionsByDatapath.get(dpId);
+
+        if (group == null) {
+            log.error("Switch {} is not registered", dpId);
+            return;
+        }
+
+        group.handleResponse(message);
+    }
+
+    /**
+     * Register switch as available for OF sessions.
+     *
+     * <p>Called from main FL thread, that why we can avoid `synchronization` here.
+     */
+    void switchActivate(DatapathId dpId) {
+        SwitchSessions group = new SwitchSessions();
+        SwitchSessions previous;
+        previous = sessionsByDatapath.put(dpId, group);
+
+        if (previous != null) {
+            log.error("Switch {} already registered (connect/disconnect race condition?)", dpId);
```

Comment:
Should "previous.disconnect();" be called to complete all bound CompletableFutures?